### ISSUE TITLE
[BUG] Amethyst window hang there in some conditions and can not recover

### ISSUE THREAD

**Issue by devfans** [amethyst/amethyst#1927]
# Tasks
- reproduce the issue
- let us know in the comments that you have the issue
- figure out what the cause is

## Description

In some conditions, the Amethyst game window will hang there can can't recover:

I am seeing one problem: I am using terminal in full screen mode, when I run the  build from terminal it'll launch the window in main screen, then after I switch back to the terminal from the main screen, I would see 
```no frame is available```
 and the amethyst window would block there.

The issue appears too if I switch the Amethyst window to full screen mode.

## Reproduction Steps

1. Launch an amethyst game.
2. Switch to the window to full screen mode.

## What You Expected to Happen

Amethyst window wont hang there

## What Actually Happened

Amethyst window hang there and can not recover

## Screenshots and Log
```
[INFO][gfx_backend_metal::device] Entry point EntryPoint { name: "main", execution_model: Vertex, work_group_size: WorkGroupSize { x: 0, y: 0, z: 0 } }
[INFO][gfx_backend_metal::device] Entry point EntryPoint { name: "main", execution_model: Fragment, work_group_size: WorkGroupSize { x: 0, y: 0, z: 0 } }
[INFO][gfx_backend_metal::device] Entry point EntryPoint { name: "main", execution_model: Vertex, work_group_size: WorkGroupSize { x: 0, y: 0, z: 0 } }
[INFO][gfx_backend_metal::device] Entry point EntryPoint { name: "main", execution_model: Fragment, work_group_size: WorkGroupSize { x: 0, y: 0, z: 0 } }
Assets loaded, swapping stats
[WARN][gfx_backend_metal::window] Swapchain drawables are changed, unable to wait for 0
[WARN][gfx_backend_metal::window] Failed to get the drawable of frame 0

[WARN][gfx_backend_metal::window] Swapchain drawables are changed, unable to wait for 1
[WARN][gfx_backend_metal::window] Failed to get the drawable of frame 1
[WARN][gfx_backend_metal::window] Swapchain drawables are changed, unable to wait for 2
[WARN][gfx_backend_metal::window] Failed to get the drawable of frame 2
[WARN][gfx_backend_metal::window] No frame is available
[WARN][gfx_backend_metal::window] No frame is available
[WARN][gfx_backend_metal::window] No frame is available
[WARN][gfx_backend_metal::window] No frame is available
[WARN][gfx_backend_metal::window] No frame is available
[WARN][gfx_backend_metal::window] No frame is available
[WARN][gfx_backend_metal::window] No frame is available
[WARN][gfx_backend_metal::window] No frame is available
```
## Your Environment
1. MacOS
2. Amethyst master branch

## Additional Context

My project: https://github.com/devfans/Godswood

**Comment by bonsairobo** [comment]
I also experience this. It only happens to me on Mac. I consider this high priority, because it happens >50% of the time, and it really slows down development.

**Comment by BedfordWest** [comment]
Confirmed this happens for me, as well. I can also confirm it only seems to happen when I launch from fullscreen terminal and not when I launch from windowed terminal. If it helps, I'm on MacOS v10.14.6 Mojave.

**Comment by jchitel** [comment]
Also happens with me. It happens around 50% of the time when I keep the window in focus and 100% of the time when the window loses focus.

**Comment by jojolepro** [comment]
Bump

**Comment by lmcgrath** [comment]
This happens to me on all examples on 0.14.0 and master as well as anything I'm writing. The pong example only shows a pair of paddles and no ball, yet the command line will show that the game is registering the ball moving and making scores. I can't make any progress with my own work because I get zero feedback graphically 100% of the time.
[log.txt](https://github.com/amethyst/amethyst/files/4177224/log.txt)

**Comment by stijnfrishert** [comment]
Happens here too, when the window loses focus.

**Comment by ea2305** [comment]
Same here, I'm getting this error on MacOS 10.15.3. In my case, When I change the window from code to the game window, the application stops to detect the pong controls (a,s,d,w and arrows).
In I got the same error "[WARN][gfx_backend_metal::window] No frame is available"

**Comment by plorefice** [comment]
The bug is still present on 0.15.0 + MacOS 10.15.4. To reproduce, simply run the application from a fullscreen terminal.

Interestingly, if I'm fast enough in switching from the fullscreen terminal to the Desktop screen (eg. by swiping on the trackpad on a MacBook Pro) everything works correctly, which seems to confirms it's somehow related to window focus.

**Comment by Neo-Zhixing** [comment]
My experience is that the program works just fine even if the window loses focus. However, the window will hang as soon as another window was dragged on top of the amethyst window.

**Comment by Spaceface16518** [comment]
@Neo-Zhixing I haven’t tested it in a while, so I can’t give any details, but I remember that happening to me a while ago. However, before that, I noticed my games CPU usage significantly decreased when I dragged another window on top of it, or was on another desktop (basically whenever there was something covering where my game was rendering). Maybe this has something to do with the way Metal behaves when the render window is hidden?

**Comment by tsal** [comment]
This has been called out as related to using an old versions of gfx / gfx-hal - I am not sure what versions to pin here to try and work around it.  Anyone had any luck?

**Comment by TransmissionsDev** [comment]
Having this as well

**Comment by CleanCut** [comment]
@jojolepro I removed the `0.15` milestone as this didn't ship in that version. My guess is that this gets addressed in #2322
